### PR TITLE
Sparse Tallies

### PR DESCRIPTION
**Pull request by wbinventor**
This PR makes it possible to store tally data in a sparse representation for data post-processing with the Python API. Sparse tallies is a first stab at making it tractable to process several GB tally datasets, and can serve as a starting point as we learn how to scale up to process TB+ datasets. This PR builds upon #525 and will be easiest to review after that has been merged. I'm not sure who is best suited to review this, but I'll let @paulromano @samuelshaner @smharper @nelsonag fight it out :-)

The sparse tally feature leverages the `scipy.sparse` module to store tally data as a sparse array if the user choose to do so. The `Tally` class now has a `sparse` property getter and setter. By default `Tally.sparse` is equal to `False` and the tally data is stored as an `ndarray`. If the user sets `Tally.sparse` equal to `True` then the underlying data arrays are converted to `lil_matrix` objects. Furthermore, since tally data is loaded from HDF5 on-the-fly as needed, any arrays which are subsequently loaded into a sparse `Tally` will be stored as `lil_matrix` objects. Of course one can always "densify" a sparse `Tally` to use `ndarray` rather than `lil_matrix` storage by simply setting `Tally.sparse` equal to `False`.

 The `Tally` class's data accessors - *e.g.*, the `sum`, `sum_sq`, `mean`, `std_dev` properties and the `get_values(...)` method - will *always* return an dense `ndarray` of the data irregardless of the underlying sparse/dense data structures. Although this was not my first choice, it seemed like the simplest option at hand, though I'm open to alternative ideas (the purpose of opening this PR is to start a discussion on the issue of large scale tally data processing more than anything).

In particular, much of the existing functionality - such as tally arithmetic, Pandas DataFrames and the test suite - relies on the use of dense `ndarrays`. Unfortunately, `scipy.sparse` does not support arrays with more than 2 dimensions, and there is a mature Python packages for general sparse array storage (though something like [DistArray](http://docs.enthought.com/distarray/ might be helpful in the future). Our tally data arrays are stored as 3D `ndarrays` which cannot be broadcasted onto 2D sparse matrices in tally arithmetic. Hence, it seemed most straightforward to simply convert the underlying sparse `lil_matrix` to a dense `ndarray` when accessed to enable the existing features to continue to work as implemented.

The `openmc.mgxs` module fully supports sparse tallies as well. The new `MGXS.sparse` and `Library.sparse` getter and setter properties leverage the `Tally.sparse` properties. This is particularly useful for energy transfer matrices, such as those needed to compute multi-group scattering matrices.

### CODE REVIEW

File: openmc/tallies.py [hunk 1]
```diff
@@ -243,6 +252,10 @@ def scores(self):
         return self._scores
 
     @property
+    def shape(self):
+        return (self.num_filter_bins, self.num_nuclides, self.num_score_bins)
```
Comment (paulromano) [inline openmc/tallies.py]: Need to add description of shape property in class docstring

File: openmc/tallies.py [hunk 2]
```diff
@@ -304,24 +317,30 @@ def nonzero(val):
...
 
             # Reshape the results arrays
-            new_shape = (nonzero(self.num_filter_bins),
-                         nonzero(self.num_nuclides),
-                         nonzero(self.num_score_bins))
-
-            sum = np.reshape(sum, new_shape)
-            sum_sq = np.reshape(sum_sq, new_shape)
+            sum = np.reshape(sum, self.shape)
+            sum_sq = np.reshape(sum_sq, self.shape)
```
Comment (paulromano) [inline openmc/tallies.py]: No need for nonzero anymore? Looks like it's unused now.

Comment (wbinventor) [inline openmc/tallies.py]: These are changes pulled in from @samuelshaner's PR, so perhaps he can
remove nonzero there. I'm waiting to fully review that PR a second time
till he gives the go ahead. It does a good job generalizing tally
arithmetic, and you may want to review it as well since the changes are
discussuon-worthy. In any case, most of the diff here will be removed once
we merge in that PR.

Comment (paulromano) [inline openmc/tallies.py]: I'm pretty sure I'm looking at the correct changeset (accounting for the fact this is based off of #525) and it still shows these as being changed in your PR only. I also explicitly looked at #525 and don't see these changes there.

Comment (wbinventor) [inline openmc/tallies.py]: Oh I see, yes this particular change is my own. I'll remove `nonzero` here. But much of the diff below is from #525 and not my own code changes for this PR.

File: openmc/tallies.py (hunk 1 above)
Comment (wbinventor) [inline openmc/tallies.py]: Good call, I always forget these things when late night coding

**Comment by paulromano** [comment]
Any particular reason you chose LIL over another format (e.g. CSR)?

**Comment by wbinventor** [comment]
Originally I chose to use LIL since i was hoping to find a way to use the
sparse representation in tally arithmetic. This would allow sparse tallies
to not only be a memory optimization, but also eliminate unnecessary flops.

At first sight LIL seemed like the most promising choice since SciPy's
implementation supports some Numpy-like fancy indexing and slicing needed
for fast tally arithmetic. However, since we have to go from 3D to 2D for
the sparse matrix storage, and elementwise operations with SciPy's sparse
matrices looks ugly, we can't easily do tally arithmetic with sparse
matrices.

As a result, there is no reason to use LIL over any other format. I'm open
to suggestions, it would even be easy to support multiple formats if we
wish. Of course we could implement our own n-dimensional sparse array API,
but I was too lazy to venture down that path.

**Comment by wbinventor** [comment]
This PR is now up-to-date and ready for review.

File: openmc/opencg_compatible.py [hunk 3]
```diff
@@ -901,7 +901,9 @@ def get_opencg_lattice(openmc_lattice):
 
     # Convert 2D universes array to 3D for OpenCG
     if len(universes.shape) == 2:
-        universes.shape = (1,) + universes.shape
+        new_universes = universes.copy()
+        new_universes.shape = (1,) + universes.shape
+        universes = new_universes
```
Comment (paulromano) [inline openmc/opencg_compatible.py]: Looks unrelated to sparsification. Can you comment?

File: openmc/tallies.py [hunk 4]
```diff
@@ -1620,7 +1701,6 @@ def hybrid_product(self, other, binary_op, filter_product=None,
     def _align_tally_data(self, other, filter_product, nuclide_product,
                           score_product):
         """Aligns data from two tallies for tally arithmetic.
-
```
Comment (paulromano) [inline openmc/tallies.py]: Add space back in

File: openmc/tallies.py [hunk 5]
```diff
@@ -1629,7 +1709,6 @@ def _align_tally_data(self, other, filter_product, nuclide_product,
         'tile' and 'repeat' operations to the new data arrays such that all
         possible combinations of the data in each tally's bins will be made
         when the arithmetic operation is applied to the arrays.
-
```
Comment (paulromano) [inline openmc/tallies.py]: Add space back in

File: openmc/tallies.py [hunk 6]
```diff
@@ -1643,18 +1722,18 @@ def _align_tally_data(self, other, filter_product, nuclide_product,
         score_product : {'tensor', 'entrywise'}
             The type of product (tensor or entrywise) to be performed between
             score data.
-
```
Comment (paulromano) [inline openmc/tallies.py]: Add space back in

File: openmc/tallies.py [hunk 7]
```diff
@@ -1643,18 +1722,18 @@ def _align_tally_data(self, other, filter_product, nuclide_product,
...
         Returns
         -------
         dict
             A dictionary of dictionaries to "aligned" 'mean' and 'std. dev'
             NumPy arrays for each tally's data.
-
```
Comment (paulromano) [inline openmc/tallies.py]: Add space back in

File: openmc/universe.py [hunk 8]
```diff
@@ -1081,7 +1081,7 @@ def get_cell_instance(self, path, distribcell_index):
         # For 2D Lattices
         if len(self._dimension) == 2:
             offset = self._offsets[i[1]-1, i[2]-1, 0, distribcell_index-1]
-            offset += self._universes[i[1]][i[2]].get_cell_instance(path,
+            offset += self._universes[i[1]-1][i[2]-1].get_cell_instance(path,
```
Comment (paulromano) [inline openmc/universe.py]: Also seemingly unrelated to sparsification. Comment?

File: openmc/tallies.py [hunk 9]
```diff
@@ -266,6 +277,10 @@ def num_bins(self):
         return num_bins
 
     @property
+    def shape(self):
+        return (self.num_filter_bins, self.num_nuclides, self.num_scores)
```
Comment (paulromano) [inline openmc/tallies.py]: Don't know why we didn't have this before! Very nice.

File: openmc/opencg_compatible.py (hunk 3 above)
Comment (wbinventor) [inline openmc/opencg_compatible.py]: You are correct! When I was updating this last night, I discovered that one of the cells in the Pandas DataFrame Jupyter Notebook example failed when building a DF for a distribcell tally using summary info from OpenCG. In particular, the code here turned the `universes` for a 2D lattice was converted into a 3D NumPy array, while leaving all of other lattice attributes such as `dimension`, `pitch` and `lower_left` as 2D.

This bug was related to PR #539 (and perhaps #534), although I found it strange since I separately ran all Notebooks while reviewing both PRs. I can submit a separate PR specifically for this bug fix, or can leave it here if we want to move this PR into the code quickly.

File: openmc/universe.py (hunk 8 above)
Comment (wbinventor) [inline openmc/universe.py]: Correct - this code change is directly related to the bug fix I elaborated in my earlier comment.

**Comment by wbinventor** [comment]
I'm not sure why I removed the blank lines in the docstrings, but I just added them back in per your comments @paulromano.

**Comment by paulromano** [comment]
One more thing -- can you add scipy to the "install_requires" in setup.py?

**Comment by wbinventor** [comment]
Sure, just added it to "setup.py". If you prefer, we could import `scipy` local by method / property as is done with `pandas` in `Tally.get_pandas_dataframe(...)`.

**Comment by paulromano** [comment]
Good point. Let's import scipy by method/property so it's not a hard requirement. You can take it out of install_requires and instead add an entry in the extras_require dictionary with 'sparse': 'scipy'.

**Comment by wbinventor** [comment]
SciPy is now an optional import.

**Comment by paulromano** [comment]
Thanks @wbinventor!
